reconcile/utils/pagerduty_api.py

```diff
@@ -1,40 +1,90 @@
-import datetime
+import logging
+from collections.abc import Iterable
+from datetime import datetime as dt
+from typing import Callable
+from typing import Optional
+from typing import Protocol
 
-import requests
 import pypd
+import requests
+from pydantic import BaseModel
+from sretoolbox.utils import retry
 
 from reconcile.utils.secret_reader import SecretReader
+from reconcile.utils.secret_reader import SupportsSecret
+
+
+class PagerDutyTargetException(Exception):
+    pass
+
+
+class PagerDutyInstance(Protocol):
+    """This protocol defines mandatory attributes/methods
+    which must be implemented by a class to be compatible."""
+
+    name: str
+
+    @property
+    def token(self) -> SupportsSecret:
+        pass
+
+
+class PagerDutyUser(Protocol):
+    """This protocol defines mandatory attributes/methods
+    which must be implemented by a class to be compatible."""
+
+    org_username: str
+    pagerduty_username: Optional[str]
+
+
+class PagerDutyTarget(Protocol):
+    """This protocol defines mandatory attributes/methods
+    which must be implemented by a class to be compatible."""
+
+    name: str
+    instance: PagerDutyInstance
+    escalation_policy_id: Optional[str]
+    schedule_id: Optional[str]
+
+
+class PagerDutyConfig(BaseModel):
+    """This protocol defines mandatory attributes/methods
+    which must be implemented by a class to be compatible."""
```

Comment:
did you mean this class to be a Protocol then?

Reply:
Not at all. The `Protocol`s are a kind of interface or blueprint without any logic. A class that implements a protocol must have all methods and attributes implemented. e.g.,

```python
# library.oy
class User(Protocol):
    username: str
    email: str


def get_username(user: User) -> str:
    return u.usename


# integration-github.py
from library import get_username

class GithubUser(BaseModel): # maybe a GQL class or something else
    username: str
    email: str
    repos: list[Repo]
    settings: list[Settings]

gh = GithubUser(username="billythekid", email="billythekid@newyork.us", repos=[], settings=[])

assert get_username(user=gh) == "billythekid" 

# integration-githlab.py
from library import get_username

class GitlabUser(BaseModel): # maybe a GQL class or something else
    username: str
    email: str
    repos: list[Repo]
    settings: list[Settings]

gl = GitlabUser(username="billythekid", email="billythekid@newyork.us", repos=[], settings=[])

assert get_username(user=gl) == "billythekid" 
```

as you can see, 2 different User classes `GithubUser` and `GitlabUser` and both can be used for `get_username` because both have `username` and `email` attributes implemented.

Reply:
I was just meaning that the docstring suggested that `PagerDutyConfig` is a protocol whereas the class inherits from `BaseModel`.

But thx a lot for the example ❤️

Reply:
Fixed the doc: 🤦